src/app/AuthenticationControllerContainer.tsx

```diff
@@ -0,0 +1,144 @@
+import InitializingScreen from './InitializingScreen';
+import { isKioskMode } from '../utils/SearchParamUtils';
+import * as MessageCenter from '../utils/MessageCenter';
+import { setServerConfig } from '../config/serverConfig';
+
+interface AuthenticationFlowProps {
+  authenticated: boolean;
+  protectedAreaComponent: React.ReactNode;
+  publicAreaComponent: React.ReactNode;
+  setGrafanaInfo: (grafanaInfo: GrafanaInfo) => void;
+  setServerStatus: (serverStatus: ServerStatus) => void;
+}
+
+interface AuthenticationFlowState {
+  stage: 'login' | 'post-login' | 'logged-in';
+  isPostLoginError: boolean;
+}
+
+class AuthenticationController extends React.Component<AuthenticationFlowProps, AuthenticationFlowState> {
+  static readonly PostLoginErrorMsg =
+    'You are logged in, but there was a problem when fetching some required server ' +
+    'configurations. Please, try refreshing the page.';
+
+  constructor(props: AuthenticationFlowProps) {
+    super(props);
+    this.state = {
+      stage: this.props.authenticated ? 'post-login' : 'login',
+      isPostLoginError: false
+    };
+  }
+
+  componentDidMount(): void {
+    if (this.state.stage === 'post-login') {
+      this.doPostLoginActions();
+    }
+
+    this.setDocLayout();
+  }
+
+  componentDidUpdate(prevProps: Readonly<AuthenticationFlowProps>, prevState: Readonly<AuthenticationFlowState>): void {
+    if (!prevProps.authenticated && this.props.authenticated) {
+      this.setState({ stage: 'post-login' });
+      this.doPostLoginActions();
+    } else if (prevProps.authenticated && !this.props.authenticated) {
+      this.setState({ stage: 'login' });
+    }
+
+    this.setDocLayout();
+  }
+
+  render() {
+    if (this.state.stage === 'logged-in') {
+      return this.props.protectedAreaComponent;
+    } else if (this.state.stage === 'post-login') {
+      return !this.state.isPostLoginError ? (
+        <InitializingScreen />
+      ) : (
+        <InitializingScreen errorMsg={AuthenticationController.PostLoginErrorMsg} />
+      );
+    } else {
+      return this.props.publicAreaComponent;
+    }
+  }
+
+  private doPostLoginActions = async () => {
+    try {
+      const getStatusPromise = API.getStatus()
+        .then(response => this.props.setServerStatus(response.data))
+        .catch(error => {
+          MessageCenter.add(API.getErrorMsg('Error fetching status.', error), 'default', MessageType.WARNING);
+        });
+      const getGrafanaInfoPromise = API.getGrafanaInfo()
+        .then(response => this.props.setGrafanaInfo(response.data))
+        .catch(error => {
+          MessageCenter.add(API.getErrorMsg('Error fetching Grafana Info.', error), 'default', MessageType.WARNING);
+        });
+
+      const configs = await Promise.all([API.getServerConfig(), getStatusPromise, getGrafanaInfoPromise]);
+      setServerConfig(configs[0].data);
+
+      this.setState({ stage: 'logged-in' });
+    } catch (err) {
+      this.setState({ isPostLoginError: true });
```

Comment:
While testing and opening Kiali the first time, I ran in that situation where `isPostLoginError` was true and it displayed the error message hardcoded above. I'm curious what would have been the caught `err` here. Maybe it should be logged? (console.error)